In [2]:
from transformers import GPT2LMHeadModel

# 1. Exploring the GPT-2 (124M) OpenAI checkpoint

In [3]:
model_hf = GPT2LMHeadModel.from_pretrained('gpt2')
sd_hf = model_hf.state_dict()

for k, v in sd_hf.items():
    print(k, v.shape)

transformer.wte.weight torch.Size([50257, 768])
transformer.wpe.weight torch.Size([1024, 768])
transformer.h.0.ln_1.weight torch.Size([768])
transformer.h.0.ln_1.bias torch.Size([768])
transformer.h.0.attn.c_attn.weight torch.Size([768, 2304])
transformer.h.0.attn.c_attn.bias torch.Size([2304])
transformer.h.0.attn.c_proj.weight torch.Size([768, 768])
transformer.h.0.attn.c_proj.bias torch.Size([768])
transformer.h.0.ln_2.weight torch.Size([768])
transformer.h.0.ln_2.bias torch.Size([768])
transformer.h.0.mlp.c_fc.weight torch.Size([768, 3072])
transformer.h.0.mlp.c_fc.bias torch.Size([3072])
transformer.h.0.mlp.c_proj.weight torch.Size([3072, 768])
transformer.h.0.mlp.c_proj.bias torch.Size([768])
transformer.h.1.ln_1.weight torch.Size([768])
transformer.h.1.ln_1.bias torch.Size([768])
transformer.h.1.attn.c_attn.weight torch.Size([768, 2304])
transformer.h.1.attn.c_attn.bias torch.Size([2304])
transformer.h.1.attn.c_proj.weight torch.Size([768, 768])
transformer.h.1.attn.c_proj.bias 

- 00:13:47 SECTION 1: implementing the GPT-2 nn.Module
- 00:28:08 loading the huggingface/GPT-2 parameters
- 00:31:00 implementing the forward pass to get logits
- 00:33:31 sampling init, prefix tokens, tokenization
- 00:37:02 sampling loop
- 00:41:47 sample, auto-detect the device
- 00:45:50 let’s train: data batches (B,T) → logits (B,T,C)
- 00:52:53 cross entropy loss
- 00:56:42 optimization loop: overfit a single batch
- 01:02:00 data loader lite
- 01:06:14 parameter sharing wte and lm_head
- 01:13:47 model initialization: std 0.02, residual init
- 01:22:18 SECTION 2: Let’s make it fast. GPUs, mixed precision, 1000ms
- 01:28:14 Tensor Cores, timing the code, TF32 precision, 333ms
- 01:39:38 float16, gradient scalers, bfloat16, 300ms
- 01:48:15 torch.compile, Python overhead, kernel fusion, 130ms
- 02:00:18 flash attention, 96ms
- 02:06:54 nice/ugly numbers. vocab size 50257 → 50304, 93ms
- 02:14:55 SECTION 3: hyperpamaters, AdamW, gradient clipping
- 02:21:06 learning rate scheduler: warmup + cosine decay
- 02:26:21 batch size schedule, weight decay, FusedAdamW, 90ms
- 02:34:09 gradient accumulation
- 02:46:52 distributed data parallel (DDP)
- 03:10:21 datasets used in GPT-2, GPT-3, FineWeb (EDU)
- 03:23:10 validation data split, validation loss, sampling revive
- 03:28:23 evaluation: HellaSwag, starting the run
- 03:43:05 SECTION 4: results in the morning! GPT-2, GPT-3 repro
- 03:56:21 shoutout to llm.c, equivalent but faster code in raw C/CUDA
- 03:59:39 summary, phew, build-nanogpt github repo

# Sources

1. [Ground truth - Let's build GPT-2 (124M), by Andrej Karpathy](https://www.youtube.com/watch?v=l8pRSuU81PU&t)
2. [Ground repository - OpenAI GPT-2](https://github.com/openai/gpt-2)
3. [Language models are unsupervised multitask learners [GPT-2 paper], by Alec Radford; Dario Amodei; Ilya Sutskever; et al.](https://cdn.openai.com/better-language-models/language_models_are_unsupervised_multitask_learners.pdf)
4. [Language Models are Few-Shot Learners [GPT-3 paper, by Alec Radford; Dario Amodei; Ilya Sutskever; et al.]](https://arxiv.org/pdf/2005.14165)